storage/execout/file.go

```diff
@@ -2,213 +2,332 @@ package execout
+	return computeDBinFilename(c.Range.StartBlock, c.Range.ExclusiveEndBlock)
+}
+
+func (c *File) ModuleName() string {
+	return c.moduleName
+}
+
+func (fw *fileWriter) SetItem(clock *pbsubstreams.Clock, data []byte) error {
 	cp := make([]byte, len(data))
 	copy(cp, data)
 
-	ci := &pboutput.Item{
+	item := &pboutput.Item{
 		BlockNum:  clock.Number,
 		BlockId:   clock.Id,
 		Timestamp: clock.Timestamp,
-		// TODO(abourget): remove the `Cursor` from this `pboutput.Item` struct,
-		//  as we're only going to store irreversible stuff now.
-		Payload: cp,
+		Payload:   cp,
 	}
 
-	if c.writingFile != nil {
-		// if we are writing the file, we flush the data and delete what was there before.
-		// if we are not writing the file, we probably writing an index so we can't delete the data, it will be aggregated at the end of the segment
-		if err := c.writePreviousKV(); err != nil {
-			return err
+	if !fw.orderedFlagWritten {
+		if _, err := streamproto.WriteOrderedBool(fw.writer); err != nil {
+			return fmt.Errorf("writing ordered bool: %w", err)
 		}
-		c.Kv = make(map[string]*pboutput.Item) // in writable File, we delete previous items
-		c.deletedBefore = &clock.Number
+		fw.orderedFlagWritten = true
 	}
 
-	c.Kv[clock.Id] = ci
+	if _, err := streamproto.WriteItem(fw.writer, item); err != nil {
+		return err
+	}
 
+	fw.lastWrittenItem = item
 	return nil
 }
 
-func (c *File) writePreviousKV() error {
-	for _, item := range c.Kv {
-		size, err := streamproto.WriteItem(c.writingFile, item)
-		if err != nil {
-			return err
+func (fr *fileReader) Get(ctx context.Context, blockNumber uint64) (payload []byte, found bool, err error) {
+	next := fr.lastReadItem
+
+	if next == nil && fr.complete { // file has no data
+		return nil, false, nil
+	}
+	for {
+		if next != nil {
+			switch {
+			case next.BlockNum == blockNumber:
+				return next.Payload, true, nil
+			case next.BlockNum > blockNumber:
+				return nil, false, nil
+			}
+		}
+		next, err = fr.ReadNext()
+		if err == io.EOF {
+			return nil, false, nil
 		}
-		c.sizeInMemory += size
-		if c.sizeInMemory > MaxExecoutSegmentSize {
-			return fmt.Errorf("%w: file %s on module %s", ErrSegmentSizeExceeded, c.Filename(), c.ModuleName)
+		if err != nil {
+			return nil, false, err
 		}
+
 	}
-	return nil
 }
 
-func (c *File) Get(clock *pbsubstreams.Clock) ([]byte, bool) {
-	c.Lock()
-	defer c.Unlock()
-	if c.deletedBefore != nil && clock.Number < *c.deletedBefore {
-		panic("trying to get deleted data, this should never happen")
+func (fr *fileReader) open(ctx context.Context) error {
+	r, err := fr.store.OpenObject(ctx, fr.Filename())
+	if err != nil {
+		return err
 	}
+	ordered, readBytes, err := streamproto.ReadOrderedBool(r)
+	if err != nil {
+		return err
+	}
+	if !ordered {
+		if err := rewriteAsOrdered(ctx, r, readBytes, fr.store, fr.ModuleName(), fr.Filename(), fr.Range, fr.logger); err != nil {
+			return err
+		}
+		if err := r.Close(); err != nil {
+			return err
+		}
```

Comment:
Does `rewriteAsOrdered` will write back to the same file that was read? I think then that `r.Close()` should be closed before attempting the write.

On local disk for example, I think the write would always fail file would be busy.